jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/rabit/RabitTracker.scala

```diff
@@ -0,0 +1,193 @@
+      .asInstanceOf[Future[Map[String, String]]]
+  ) match {
+    case Success(fut) => fut
+    case Failure(e) =>
+      val delayedFailure = Promise[Map[String, String]]()
+      delayedFailure.failure(e)
+      delayedFailure.future
+  }
+
+  // a future for XGBoost worker execution
+  val futureCompleted: Future[Int] = Try(
+    Await.result(handler ? RabitTrackerHandler.RequestCompletionFuture, askTimeout.duration)
+      .asInstanceOf[Future[Int]]
+  ) match {
+    case Success(fut) => fut
+    case Failure(_) =>
+      val delayedFailure = Promise[Int]()
+      // mark 0 worker as completed, will trigger AssertionError
+      // when call waitFor()
+      delayedFailure.success(0)
+      delayedFailure.future
+  }
+
+  /**
+    * Start the Rabit tracker.
+    *
+    * @param timeout The timeout for awaiting connections from worker nodes.
+    *        Note that when used in Spark applications, because all Spark transformations are
+    *        lazily executed, the I/O time for loading RDDs/DataFrames from external sources
+    *        (local dist, HDFS, S3 etc.) must be taken into account for the timeout value.
+    *        If the timeout value is too small, the Rabit tracker will likely timeout before workers
+    *        establishing connections to the tracker, due to the overhead of loading data.
+    *        Using a finite timeout is encouraged, as it prevents the tracker (thus the Spark driver
+    *        running it) from hanging indefinitely due to worker connection issues (e.g. firewall.)
+    * @return Boolean flag indicating if the Rabit tracker starts successfully.
+    */
+  def start(timeout: Duration = 10 minutes): Boolean = {
+    handler ? RabitTrackerHandler.StartTracker(
+      new InetSocketAddress(InetAddress.getLocalHost, port.getOrElse(0)), maxPortTrials, timeout)
+
+    // The success of the Future is contingent on binding to an InetSocketAddress.
+    Try(Await.ready(futureWorkerEnvs, tcpBindTimeout)).isSuccess
+  }
+
+  /**
+    * Start the Rabit tracker.
+    *
+    * @param connectionTimeoutMillis Timeout, in milliseconds, for the tracker to wait for worker
+    *                                connections. If a non-positive value is provided, the tracker
+    *                                waits for incoming worker connections indefinitely.
+    * @return Boolean flag indicating if the Rabit tracker starts successfully.
+    */
+  def start(connectionTimeoutMillis: Long): Boolean = {
+    if (connectionTimeoutMillis <= 0) {
+      start(Duration.Inf)
+    } else {
+      start(Duration.fromNanos(connectionTimeoutMillis * 1e6))
+    }
+  }
+
+  /**
+    * Get a Map of necessary environment variables to initiate Rabit workers.
+    *
+    * @return HashMap containing tracker information.
+    */
+  def getWorkerEnvs: java.util.Map[String, String] = {
+    new java.util.HashMap((Await.result(futureWorkerEnvs, 0 nano) ++ Map(
```

Comment:
The Future must have already succeeded by the time the user calls `getWorkerEnvs`, therefore there's no need to have any timeout at all. 

To echo your comment above, probably this Future is not needed.